### DISCUSSION TITLE
Implement consent for analytics

### DISCUSSION THREAD

**Discussion by mikehardy** [invertase/react-native-firebase#6636] [Ideas] [answered]
Would love to see a PR that implements all the consent features for analytics.

I believe it consists of:

- exposing the analytics storage default in AndroidManifest ad Apple plist:
  - https://developers.google.com/tag-platform/devguides/app-consent?platform=android#analytics_storage
  - https://developers.google.com/tag-platform/devguides/app-consent?platform=ios#analytics_storage

- exposing the ad storage default in AndroidManifest and iOS plist:
  - https://developers.google.com/tag-platform/devguides/app-consent?platform=android#ad_storage
  -   - https://developers.google.com/tag-platform/devguides/app-consent?platform=ios#ad_storage

There are similar PRs exposing other AndroidManifest attributes that anyone may reference by using git blame on the firebase.json file - https://github.com/invertase/react-native-firebase/blame/main/tests/firebase.json

Expose the native platform APIs that implement consent setting:

https://developers.google.com/tag-platform/devguides/app-consent?platform=android#update_consent
https://developers.google.com/tag-platform/devguides/app-consent?platform=ios#update_consent

API reference for firebase-js-sdk setConsent:
- https://firebase.google.com/docs/reference/js/analytics.md#setconsent
  - https://firebase.google.com/docs/reference/js/analytics.consentsettings.md#consentsettings_interface (only ad and analytics apply as the others are not implemented in native SDKs, ignore rest I suppose)

This is related to analytics session ID and app instance id, as it will return null if consent is denied for analytics_storage
- https://firebase.google.com/docs/reference/android/com/google/firebase/analytics/FirebaseAnalytics#public-taskstring-getappinstanceid (retrieves the app instance id from the service or `null` if analytics storage ha sbeen set to denied)
- https://firebase.google.com/docs/reference/android/com/google/firebase/analytics/FirebaseAnalytics#public-tasklong-getsessionid (likewise, returns null if analytics storage set to denied - *or* if session is expired - use setSessionTimeOutInterval to force session to expire and check null return? use resetAnalyticsData to force session to expire after known wait to check null?)

----

Already completed:

~~Also should expose "resetAnalyticsData" (which may serve to expire the session, allowing for a test of null session return even if consent is set to true...~~

- ~~https://firebase.google.com/docs/reference/android/com/google/firebase/analytics/FirebaseAnalytics#public-void-resetanalyticsdata~~
- ~~https://firebase.google.com/docs/reference/ios/firebaseanalytics/api/reference/Classes/FIRAnalytics#+resetanalyticsdata~~

**Comment by dlackty** [comment]
Good news is that `resetAnalyticsData` has already been implemented in `react-native-firebase` 😉 

https://github.com/invertase/react-native-firebase/blob/72401288eb7cbfe3246ffaa384985135a74c929a/packages/analytics/lib/index.d.ts#L763-L773

**Reply by mikehardy** [reply]
Haha - nice find, you think I would know about it :see_no_evil: - this issue was mostly just writing down all my upstream docs research at once. This comment points out clearly that it needs a second pass filtering for what exists. Thanks @dlackty !

**Comment by tnke** [comment]
@mikehardy Any updates on the `setConsent` support? I'm in a situation where I'd need to implement Google's Consent Mode to a React Native app. Although don't have experience on building native library support for RN, I could try to give it a go.

**Reply by mikehardy** [reply]
Hey @tnke no progress on this so far, and it isn't on the current priority list. However, something that is *always* on my priority list is helping collaborate + shepherd any community proposed PRs into the library so if you gave it a go and posted a PR I'd be happy to help get a PR merged

**Comment by Julian-Lundberg** [comment]
@mikehardy Hey Mike! Am currently responsible for developing a cookie "banner" in an app, and I am in dire need to actually get access to changing consents. EU is just now in the process of implementing requirements for granular control of consents in apps. It's pretty urgent too, my deadline is end of february for this. I am in the same boat as @tnke, where I have no experience on building on this library, and of course I could give it a try as well. Did you make any progress @tnke ?

**Reply by tnke** [reply]
@Julian-Lundberg I haven't had the chance to progress this yet.

**Comment by SimpleCreations** [comment]
In case anyone finds this useful, here's a patch I created that adds the missing functionality. (Unfortunately I didn't have time to prepare a full PR.)

https://gist.github.com/SimpleCreations/26ed8acc3f67f22188bd65fa74056ab7

In addition to applying this patch, you'd also have to add the default consent values to AndroidManifest.xml and Info.plist, as described in the [official guide](https://developers.google.com/tag-platform/security/guides/app-consent).

@tnke @Julian-Lundberg 👀

**Reply by Julian-Lundberg** [reply]
Thats great! Thank you for sharing your approach :)

**Reply by Summys** [reply]
Thanks @SimpleCreations. I used it to create a PR https://github.com/invertase/react-native-firebase/pull/7629/files 
@mikehardy If you can review it, that would be great. These changes from what I understand have to be applied by 6th March for EEA users. It's my first contribution to an open source project, so bear with me.

**Reply by mikehardy** [reply]
Fantastic, I see it in the queue, sincere thanks @SimpleCreations and @Summys

**Reply by nixolas1** [reply]
Tested the newest patch from the PR out, and see no errors! (Using Expo 50, patching 18.8.0, adding defaults in firebase.json etc)

EDIT: 
following the official guide and [setting](https://firebase.google.com/docs/analytics/events?platform=android#view_events_in_the_android_studio_debug_log) the logcat verbosity level higher, it seems like the JS consent trigger shows up!

`02-20 10:51:19.568 V/FA-SVC  (20981): Setting consent, package, consent: com.fake.app.id, settings: source=100, AD_STORAGE=granted, ANALYTICS_STORAGE=granted`

I also see this prop being set in the actual event bundle that is uploaded, not sure if it means all the consents are true.

`consent_signals: G111`

But I can't see the new v2 consents showing up anywhere. AD_PERSONALIZATION and AD_USER_DATA. Not sure if this is just because of the native package just not logging those options, or if there is something actually wrong with which consents are being set.

**Reply by mikehardy** [reply]
Thank you for testing! I'm sorry I don't have an answer to your questions on the v2 consents, but knowing the patch is at least *partially* working is very good confirmation to hear. Unfortunately  the only thing I can think of for more verification is a kind of arduous step of temporarily integrating react-native-google-mobile-ads (which I help maintain) so you can use the methods they export for user consent checking https://docs.page/invertase/react-native-google-mobile-ads/european-user-consent#inspecting-consent-choices - and big warning - I'm not sure if these things are 100% compatible in how they are storing and exposing what they are exposing. Is it the *same* data store they are all reading/writing to, so that inspecting AdsConsent.getUserChoices() from react-native-google-mobile-ads will show what the user selected here? I cannot really say 

At least to reduce the arduousness of maybe trying it down to mostly just adding the module, you can re-use most of the example app / test identifier stuff from the repo's sample https://github.com/invertase/react-native-google-mobile-ads/blob/main/RNGoogleMobileAdsExample/app.json

**Reply by Summys** [reply]
I confirm that I get the same output on Android as @nixolas1 
gcs: G111 means the consent is given to Google Ads and Google Analytics. Which is basic mode. 
In Advanced mode we should be able to see a "gcd" parameter, which I don't get on my end. 

The upgrade seems quite straightforward and followed it step by step
https://developers.google.com/tag-platform/security/guides/app-consent?consentmode=advanced&platform=android#upgrade-consent-v2

**Comment by mikehardy** [comment, marked as answer]
releasing now, should pop out as 18.9.0

**Reply by Julian-Lundberg** [reply]
Amazing! thank all of you!

**Comment by aleks-khomenko** [comment]
Hello guys :grinning:
First of all I want to say: "Thank you :heart: for your hard work"
I wish one day I can join and help. Currently I am learning and improving my skills :exploding_head:.
I would like to ask a few questions regarding consent mode v2.
1. Should we request the consent from user for each of these items before set the value?
`await firebase.analytics().setConsent({
  analytics_storage: true,
  ad_storage: true,
  ad_user_data: true,
  ad_personalization: true,
});`
As I understood from the previous comments  `analytics_storage`, `ad_storage` are set `true` by default.
2. If I don't use advertising in my app, can I set `ad_user_data` and `ad_personalization` to `false` and do not ask user's consent?

I will be appreciate for answers. Also if someone can describe some general steps regarding consent flow in general it will be very helpful.
